src/oscar/templatetags/category_tags.py

```diff
@@ -18,36 +79,55 @@ def get_annotated_list(depth=None, parent=None):
     max_depth = depth
 
     annotated_categories = []
+    tree_slug = ""
 
     start_depth, prev_depth = (None, None)
     if parent:
         categories = parent.get_descendants()
+        tree_slug = parent.get_full_slug()
         if max_depth is not None:
             max_depth += parent.get_depth()
     else:
         categories = Category.get_tree()
 
-    info = {}
+    if max_depth is not None:
+        categories = categories.filter(depth__lte=max_depth)
+
+    info = CheapCategoryInfo(parent, url="")
+
     for node in categories:
         node_depth = node.get_depth()
         if start_depth is None:
             start_depth = node_depth
-        if max_depth is not None and node_depth > max_depth:
-            continue
 
         # Update previous node's info
-        info['has_children'] = prev_depth is None or node_depth > prev_depth
+        if prev_depth is None or node_depth > prev_depth:
+            info["has_children"] = True
+            if info.category is not None:
+                tree_slug = info.category.get_full_slug(tree_slug)
+
         if prev_depth is not None and node_depth < prev_depth:
-            info['num_to_close'] = list(range(0, prev_depth - node_depth))
+            depth_difference = prev_depth - node_depth
+            info["num_to_close"] = list(range(0, depth_difference))
+            tree_slugs = tree_slug.rsplit(node._slug_separator, depth_difference)
+            if tree_slugs:
+                tree_slug = tree_slugs[0]
+            else:
+                tree_slug = node.slug
+
+        info = CheapCategoryInfo(
+            node,
+            url=node.get_absolute_url(tree_slug),
+            num_to_close=[],
+            level=node_depth - start_depth,
+        )
+        annotated_categories.append(info)
 
-        info = {'num_to_close': [],
-                'level': node_depth - start_depth}
-        annotated_categories.append((node, info,))
         prev_depth = node_depth
 
     if prev_depth is not None:
         # close last leaf
-        info['num_to_close'] = list(range(0, prev_depth - start_depth))
-        info['has_children'] = prev_depth > prev_depth
+        info["num_to_close"] = list(range(0, prev_depth - start_depth))
```

Comment:
Please avoid unnecessary syntax changes - this make it hard to use `blame` to track down changes in future.